File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/ItemCategorySyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;
import java.util.List;

import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.db.ItemCategory;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;
import io.requery.query.Tuple;

public class ItemCategorySyncAdapter extends BaseConditionalSyncAdapter<ItemCategory, Long> {
    public ItemCategorySyncAdapter(BlockingEntityStore<Persistable> store, FileStorage fileStorage, String eventSlug, PretixApi api, SyncManager.ProgressFeedback feedback) {
        super(store, fileStorage, eventSlug, api, feedback);
    }

    @Override
    public void updateObject(ItemCategory obj, JSONObject jsonobj) throws JSONException {
        obj.setEvent_slug(eventSlug);
        obj.setServer_id(jsonobj.getLong("id"));
        obj.setPosition(jsonobj.getLong("position"));
        obj.setIs_addon(jsonobj.optBoolean("is_addon", false));
        obj.setJson_data(jsonobj.toString());
    }

    @Override
    public Iterator<ItemCategory> runBatch(List<Long> ids) {
        return store.select(ItemCategory.class)
                .where(ItemCategory.EVENT_SLUG.eq(eventSlug))
                .and(ItemCategory.SERVER_ID.in(ids))
                .get().iterator();
    }

    @Override
    Iterator<Tuple> getKnownIDsIterator() {
        return store.select(ItemCategory.SERVER_ID)
                .where(ItemCategory.EVENT_SLUG.eq(eventSlug))
                .get().iterator();
    }

    @Override
    String getResourceName() {
        return "categories";
    }

    @Override
    Long getId(JSONObject obj) throws JSONException {
        return obj.getLong("id");
    }

    @Override
    Long getId(ItemCategory obj) {
        return obj.getServer_id();
    }

    @Override
    ItemCategory newEmptyObject() {
        return new ItemCategory();
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/ItemCategorySyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import app.cash.sqldelight.db.QueryResult
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.sqldelight.ItemCategory
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import org.json.JSONObject

class ItemCategorySyncAdapter(
    db: SyncDatabase,
    fileStorage: FileStorage,
    eventSlug: String,
    api: PretixApi,
    syncCycleId: String,
    feedback: ProgressFeedback?,
) : BaseConditionalSyncAdapter<ItemCategory, Long>(
    db = db,
    fileStorage = fileStorage,
    eventSlug = eventSlug,
    api = api,
    syncCycleId = syncCycleId,
    feedback = feedback,
) {

    override fun getResourceName(): String = "categories"

    override fun getId(obj: ItemCategory): Long = obj.server_id!!

    override fun getId(obj: JSONObject): Long = obj.getLong("id")

    override fun getJSON(obj: ItemCategory): JSONObject = JSONObject(obj.json_data)

    override fun queryKnownIDs(): MutableSet<Long>? {
        val res = mutableSetOf<Long>()
        db.itemCategoryQueries.selectServerIdsByEventSlug(event_slug = eventSlug)
            .execute { cursor ->
                while (cursor.next().value) {
                    val id = cursor.getLong(0) ?: throw RuntimeException("id column not available")
                    res.add(id)
                }

                QueryResult.Unit
            }

        return res
    }

    override fun insert(jsonobj: JSONObject) {
        db.itemCategoryQueries.insert(
            event_slug = eventSlug,
            is_addon = jsonobj.optBoolean("is_addon", false),
            json_data = jsonobj.toString(),
            position = jsonobj.getLong("position"),
            server_id = jsonobj.getLong("id"),
        )
    }

    override fun update(obj: ItemCategory, jsonobj: JSONObject) {
        db.itemCategoryQueries.updateFromJson(
            event_slug = eventSlug,
            is_addon = jsonobj.optBoolean("is_addon", false),
            json_data = jsonobj.toString(),
            position = jsonobj.getLong("position"),
            id = obj.id,
        )
    }

    override fun delete(key: Long) {
        db.itemCategoryQueries.deleteByServerId(key)
    }

    override fun runInTransaction(body: TransactionWithoutReturn.() -> Unit) {
        db.itemCategoryQueries.transaction(false, body)
    }

    override fun runBatch(parameterBatch: List<Long>): List<ItemCategory> =
        db.itemCategoryQueries.selectByServerIdListAndEventSlug(
            server_id = parameterBatch,
            event_slug = eventSlug,
        ).executeAsList()
}
